Convert URL types to strongly typed CSS/Style values

weinig opened the pull request:
#### 0edc4f619186f87c8862425f99da70e257ca8281
<pre>
Convert URL types to strongly typed CSS/Style values
<a href="https://bugs.webkit.org/show_bug.cgi?id=291045">https://bugs.webkit.org/show_bug.cgi?id=291045</a>

Reviewed by Antti Koivisto and Darin Adler.

Update parsing and representation of &lt;url&gt; productions as
defined in CSS Values and Units 4 (<a href="https://www.w3.org/TR/css-values-4/#urls)">https://www.w3.org/TR/css-values-4/#urls)</a>
to use strong value types and have consistent behavior for
serialization and style building. This fixes a few WPT test
failures.

As URLs are no longer representable by CSSPrimitiveValue,
a new CSSURLValue type is also added that wraps CSS::URL.

This change allows removing a lot of unnecessary re-resolution
that was going on in the image code, as the only re-resolution
needed can be performed by style building. Ultimately, it would
be better if we could remove that final case as well, by finding
a good hook to re-parse necessary styles adopted from documents,
but that can be done separately.

To allow using Style::URL in ReferenceFilterOperation without a
laying violation (or rather, without additional layering violations
as it was already using non-platform types), ReferenceFilterOperation
was moved out of platform. This matches what was done recently for
DropShadowFilterOperationWithStyleColor.

* LayoutTests/fast/masking/parsing-clip-path-iri-expected.txt:
* LayoutTests/fast/masking/parsing-clip-path-iri.html:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/fill-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/stroke-computed-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSCursorImageValue.cpp:
* Source/WebCore/css/CSSCursorImageValue.h:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSFilterPropertyValue.cpp:
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
* Source/WebCore/css/CSSFontFaceSrcValue.h:
* Source/WebCore/css/CSSImageValue.cpp:
* Source/WebCore/css/CSSImageValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSURLValue.cpp: Added.
* Source/WebCore/css/CSSURLValue.h: Added.
* Source/WebCore/css/CSSUnits.cpp:
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.cpp:
* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
* Source/WebCore/css/calc/CSSCalcTree+NumericIdentity.h:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcType.cpp:
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Masking.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Motion.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.h:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/typedom/CSSStyleImageValue.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/values/CSSValueTypes.cpp:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/filter-effects/CSSFilterReference.h:
* Source/WebCore/css/values/primitives/CSSURL.cpp: Added.
* Source/WebCore/css/values/primitives/CSSURL.h: Added.
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
* Source/WebCore/page/ElementTargetingController.cpp:
* Source/WebCore/platform/graphics/filters/FilterOperation.cpp:
* Source/WebCore/platform/graphics/filters/FilterOperation.h:
* Source/WebCore/rendering/CSSFilter.cpp:
* Source/WebCore/rendering/PathOperation.cpp:
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/ReferencedSVGResources.h:
* Source/WebCore/rendering/RenderLayerFilters.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/style/ReferenceFilterOperation.cpp: Added.
* Source/WebCore/rendering/style/ReferenceFilterOperation.h: Added.
* Source/WebCore/rendering/style/SVGRenderStyle.h:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.h:
* Source/WebCore/rendering/style/StyleCachedImage.cpp:
* Source/WebCore/rendering/style/StyleCachedImage.h:
* Source/WebCore/rendering/style/StyleCursorImage.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.h:
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
* Source/WebCore/rendering/style/StyleImage.h:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResources.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/filter-effects/StyleFilterProperty.cpp:
* Source/WebCore/style/values/filter-effects/StyleFilterReference.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleFilterReference.h: Added.
* Source/WebCore/style/values/primitives/StyleURL.cpp: Added.
* Source/WebCore/style/values/primitives/StyleURL.h: Added.
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.cpp:
* Source/WebCore/svg/SVGFontFaceUriElement.cpp:
* Source/WebCore/svg/SVGURIReference.cpp:
* Source/WebCore/svg/SVGURIReference.h:
* Source/WebCore/svg/graphics/SVGResourceImage.cpp:
* Source/WebCore/svg/graphics/SVGResourceImage.h:

Canonical link: <a href="https://commits.webkit.org/293295@main">https://commits.webkit.org/293295@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a74a4dbe921f0d66b50a6a0e28c6458d994b484

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98445 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18076 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8307 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103566 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/48976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100489 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18369 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26528 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/74940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/48976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101449 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/13926 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88912 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55300 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13708 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6877 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48415 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6954 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105939 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25533 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83908 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25909 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85113 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83388 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21066 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28036 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/5705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/19187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25492 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/30669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25310 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28630 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26885 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/css/values/primitives/CSSURL.cpp:
```diff
@@ -0,0 +1,95 @@
+{
+    return !url.specified.isEmpty()
+        && !url.specified.startsWith('#')
+        && !protocolIs(url.specified, "data"_s);
```
Doing scheme comparisons on a string won't work well. This won't account for inputs where the scheme contains `\t` or `\n`. It also seems very weird to special case `data:` URLs here. Is that really per specification?

weinig replied:
Not as an explanation, but this part is existing and is just moved. 

I believe this function is trying to just be conservative, hence the “may” in mayDependOnBaseURL, but I don’t know if that makes it ok. 

That said, we can use the resolved value here and check the URL protocol instead.

annevk replied:
Yeah I think the original code used `WebCoreURLInstance.protocolIsData()` when I skimmed through your patch. Doing that here as well would be preferable and maybe we should add a FIXME as this is deeply suspect.

weinig replied:
Should we rename the non-instance protocolIs to something more scary so people don’t make this mistake?

annevk replied:
Yeah that would be good. No great suggestion other than giving them an `Unsafe` suffix.

weinig replied:
Maybe we just get rid of it? Replace all existing uses of `protocolIs(string, "foo"_s)` with `URL(string).protocolIs("foo"_s)`?

weinig replied:
Hm, actually, looking at the code for protocolIs(), which just calls protocolIsInternal(), it does look like it skips tabs and newlines -> https://github.com/WebKit/WebKit/blob/e7b2eb32d8a207ffb20c38d3513c4948dd5c3ae4/Source/WTF/wtf/URL.cpp#L869

annevk replied:
Oh sorry, I must have misremembered. (The original concern about the logic of this stands, but since this is a refactoring that's not really applicable.)

darinadler commented on LayoutTests/fast/masking/parsing-clip-path-iri.html:
```diff
@@ -1,47 +1,28 @@
 <!DOCTYPE html>
 <html>
+<head>
+<script src="../../resources/testharness.js"></script>
+<script src="../../resources/testharnessreport.js"></script>
+<script src="../../imported/w3c/web-platform-tests/css/support/parsing-testcommon.js"></script>
+<script src="../../imported/w3c/web-platform-tests/css/support/computed-testcommon.js"></script>
```
Seems a little inelegant to have tests in `fast` that actually require the WPT harness. Is this the first one?

weinig replied:
There are actually quite a few.

darinadler commented on Source/WebCore/css/CSSCursorImageValue.h:
```diff
@@ -38,10 +39,11 @@ class BuilderState;
 public:
     static Ref<CSSCursorImageValue> create(Ref<CSSValue>&& imageValue, RefPtr<CSSValue>&& hotSpot, LoadedFromOpaqueSource);
-    static Ref<CSSCursorImageValue> create(Ref<CSSValue>&& imageValue, RefPtr<CSSValue>&& hotSpot, URL, LoadedFromOpaqueSource);
+    static Ref<CSSCursorImageValue> create(Ref<CSSValue>&& imageValue, RefPtr<CSSValue>&& hotSpot, CSS::URL, LoadedFromOpaqueSource);
```
Why use `Ref&&` but `CSS::URL` that is not an rvalue reference in the same constructor?

weinig replied:
Just a mistake. Fixed.

darinadler commented on Source/WebCore/css/CSSCustomPropertyValue.cpp:
```diff
@@ -32,6 +32,7 @@
 #include "CSSParserIdioms.h"
 #include "CSSSerializationContext.h"
 #include "CSSTokenizer.h"
+#include "CSSURL.h"
```
Pretty sure this is not needed. The `StyleURL.h` header has to include this so that `toCSS` can be compiled.

darinadler commented on Source/WebCore/css/CSSFontFaceSrcValue.cpp:
```diff
@@ -74,16 +74,20 @@ bool CSSFontFaceSrcLocalValue::equals(const CSSFontFaceSrcLocalValue& other) con
+    // This resolution would normally happen when converting to Style::URL, but since this
+    // is a font-face descriptor, that never happens.
+    if (m_location.resolved.isNull())
+        m_location.resolved = URL { m_location.specified };
```
I’m surprised to see this written out like this. I would think it would be a named function. In fact, the name of the function can help document further what is going on here.

weinig replied:
Yeah, that's a good idea. Will do.

darinadler commented on Source/WebCore/css/CSSValue.cpp:
```diff
@@ -369,6 +372,7 @@ Ref<DeprecatedCSSOMValue> CSSValue::createDeprecatedCSSOMWrapper(CSSStyleDeclara
     case Quad:
     case Rect:
     case ValuePair:
+    case URL:
```
Alphabetical order?

weinig replied:
Eek.

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.h:
```diff
@@ -97,7 +98,7 @@ class ComputedStyleExtractor {
 
     RefPtr<CSSValue> svgPropertyValue(CSSPropertyID) const;
-    Ref<CSSValue> adjustSVGPaint(SVGPaintType, const String& url, Ref<CSSValue> color) const;
+    Ref<CSSValue> adjustSVGPaint(SVGPaintType, const Style::URL&, Ref<CSSValue> color, const RenderStyle&) const;
```
Slightly strange and inconsistent that `color` here is a `Ref`, not `const Ref&` nor `Ref&&` nor `const CSSValue&` nor `CSSValue&`.

weinig replied:
Cleanest is just to pass the `const CSS::Color&`. Fixed.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -1181,8 +1181,11 @@ static RefPtr<CSSValue> consumeImageSet(CSSParserTokenRange& args, CSS::Property
     if (range.peek().type() == StringToken && allowedImageTypes.contains(AllowedImageType::RawStringAsURL)) {
-        return CSSImageValue::create(state.context.completeURL(range.consumeIncludingWhitespace().value().toAtomString().string()),
-            state.context.isContentOpaque ? LoadedFromOpaqueSource::Yes : LoadedFromOpaqueSource::No);
+        auto url = CSS::completeURL(range.peek().value().toAtomString().string(), state.context);
```
I go out of my way to not have local variables named `url` just because the all-lowercase acronym annoys me so much. Here I would have used `imageURL`. I’ve never had the courage to use the word `locator`, though!

weinig replied:
I was kind of considering naming the types CSS::Locator/Style::Locator or CSS::ResourceLocator/Style::ResourceLocator or CSS::Resource/Style::Resource as that is in the name of the section in the spec and these aren't really semantically the same as URL:
"4.5. Resource Locators: the `<url>` type" (https://www.w3.org/TR/css-values-4/#urls).

Might still change it at some point when adding the `<url-modifier>` parts, but for now this won out.

I can definitely avoid using lowercase though.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp:
```diff
@@ -36,35 +42,48 @@ namespace CSSPropertyParserHelpers {
 {
-    auto url = consumeURLRaw(range);
-    if (url.isNull())
+    auto url = consumeURLRaw(range, state);
```
Another example (I didn’t comment on all of them) where I would wriggle to avoid the name `url`:
```suggestion
    auto rawURL = consumeURLRaw(range, state);
```

darinadler commented on Source/WebCore/rendering/ReferencedSVGResources.cpp:
```diff
@@ -272,10 +269,11 @@ RefPtr<SVGElement> ReferencedSVGResources::referencedPaintServerElement(TreeScop
         return nullptr;
+
     RefPtr element = elementForResourceID(treeScope, referenceFilter.fragment(), SVGNames::filterTag);
     return element ? downcast<SVGFilterElement>(WTFMove(element)) : nullptr;
```
Pretty sure downcast with `RefPtr` already handles null. So this does the same thing and we should consider switching to it.
```suggestion
    return downcast<SVGFilterElement>(elementForResourceID(treeScope, referenceFilter.fragment(), SVGNames::filterTag));
```

darinadler commented on Source/WebCore/rendering/style/ReferenceFilterOperation.cpp:
```diff
@@ -0,0 +1,78 @@
+
+bool ReferenceFilterOperation::isIdentity() const
+{
+    // Answering this question requires access to the renderer and the referenced filterElement.
+    ASSERT_NOT_REACHED();
+    return false;
```
The comment indicates why we can’t accurately answer. But it doesn’t indicate why we know this won’t ever be called. I think we’d want to explain that too.

darinadler commented on Source/WebCore/rendering/style/ReferenceFilterOperation.cpp:
```diff
@@ -0,0 +1,78 @@
+
+IntOutsets ReferenceFilterOperation::outsets() const
+{
+    // Answering this question requires access to the renderer and the referenced filterElement.
+    ASSERT_NOT_REACHED();
+    return { };
```
The comment indicates why we can’t accurately answer. But it doesn’t indicate why we know this won’t ever be called. I think we’d want to explain that too.

weinig replied:
I agree, but I don't have quite enough knowledge of this code to answer it without a deeper dive. (The comment is pre-existing and just being moved along with the function).

darinadler commented on Source/WebCore/rendering/style/ReferenceFilterOperation.h:
```diff
@@ -0,0 +1,84 @@
+
+namespace Style {
+
+class ReferenceFilterOperation : public FilterOperation {
```
Can more of this class be `final`?

darinadler commented on Source/WebCore/rendering/style/SVGRenderStyleDefs.h:
```diff
@@ -186,8 +187,8 @@ class StyleStrokeData : public RefCounted<StyleStrokeData> {
 
-    String paintUri;
-    String visitedLinkPaintUri;
+    Style::URL paintUri;
+    Style::URL visitedLinkPaintUri;
```
Do you know why these names have `Uri` in them?

weinig replied:
Weird old SVG spec reasons I am sure. Not something we need to maintain.

darinadler commented on Source/WebCore/rendering/style/StyleCachedImage.cpp:
```diff
@@ -106,16 +118,14 @@ LegacyRenderSVGResourceContainer* StyleCachedImage::uncheckedRenderSVGResource(c
     if (!m_cachedImage) {
-        auto fragmentIdentifier = SVGURIReference::fragmentIdentifierFromIRIString(reresolvedURL.string(), document);
+        Ref document = renderer->document();
+        auto fragmentIdentifier = SVGURIReference::fragmentIdentifierFromIRIString(m_url, document);
```
```suggestion
        auto fragmentIdentifier = SVGURIReference::fragmentIdentifierFromIRIString(m_url, renderer->protectedDocument());
```

darinadler commented on Source/WebCore/rendering/style/StyleCachedImage.cpp:
```diff
@@ -146,7 +155,7 @@ RenderSVGResourceContainer* StyleCachedImage::renderSVGResource(const RenderElem
         return nullptr;
 
-    if (imageURL().string().find('#') == notFound)
+    if (m_url.resolved.string().find('#') == notFound)
```
```suggestion
    if (!m_url.resolved.string().contains('#'))
```
or if we aren’t worried changing behavior if the URL is invalid
```suggestion
    if (!m_url.resolved.hasFragmentIdentifier())
```

darinadler commented on Source/WebCore/rendering/style/StyleCursorImage.h:
```diff
@@ -40,7 +40,8 @@ enum class LoadedFromOpaqueSource : bool;
 public:
-    static Ref<StyleCursorImage> create(Ref<StyleImage>&&, const std::optional<IntPoint>&, const URL&, LoadedFromOpaqueSource);
+    static Ref<StyleCursorImage> create(const Ref<StyleImage>&, const std::optional<IntPoint>&, const Style::URL&, LoadedFromOpaqueSource);
+    static Ref<StyleCursorImage> create(Ref<StyleImage>&&, std::optional<IntPoint>&&, Style::URL&&, LoadedFromOpaqueSource);
```
An optional `IntPoint` doesn’t really need move semantics. It’s only 65 bits and a simple type with no side effects.
```suggestion
    static Ref<StyleCursorImage> create(Ref<StyleImage>&&, std::optional<IntPoint>, Style::URL&&, LoadedFromOpaqueSource);
```

darinadler commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -872,22 +874,25 @@ inline RefPtr<StylePathData> BuilderConverter::convertDPath(BuilderState& builde
+        auto styleURL = Style::toStyle(url->url(), builderState);
+
+        auto fragment = SVGURIReference::fragmentIdentifierFromIRIString(styleURL, builderState.document());
+        // FIXME: It doesn't work with external SVG references (see https://bugs.webkit.org/show_bug.cgi?id=126133)
```
```suggestion
        // FIXME: This doesn't work correctly for external SVG references (see https://bugs.webkit.org/show_bug.cgi?id=126133)
```
Also would love to be more specific than “doesn’t work correctly” if I knew what was wrong.

weinig replied:
This one I totally do know the answer and can improve that comment.

darinadler commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1891,15 +1896,16 @@ inline float BuilderConverter::convertOpacity(BuilderState& builderState, const
     auto* primitiveValue = requiredDowncast<CSSPrimitiveValue>(builderState, value);
     if (!primitiveValue)
-        return { };
+        return Style::URL::empty();
```
If the new code is correct then we leave this paragraph out entirely:
```suggestion
```
This relies on knowing that the change from null to empty is good or at least harmless, and that there are no side effects of `requiredDowncast`.

weinig replied:
Unfortunately, there are side effects to `requiredDowncast`. It sets `setCurrentPropertyInvalidAtComputedValueTime` on failures.

darinadler commented on Source/WebCore/style/values/primitives/StyleURL.h:
```diff
@@ -0,0 +1,65 @@
+    WTF::URL resolved;
+
+    static URL empty() { return { .resolved = { } }; }
+    bool isEmpty() const { return resolved.isNull(); }
```
There is a distinct empty value in `WTF::URL` and we’re not doing anything in the `struct` to prevent that value from getting in, and if it did then `isEmpty()` will return `false` and `resolved.isEmpty()` will return `true`. Seems a bit messy.

weinig replied:
Yeah, we actually need the distinct empty value. I will change the name of these to use `none`. In a future change I would like to have distinct types for Style::URL and Style::NoneOrURL.

darinadler commented on Source/WebCore/svg/SVGFontFaceUriElement.cpp:
```diff
@@ -58,10 +59,7 @@ SVGFontFaceUriElement::~SVGFontFaceUriElement()
-    location.specifiedURLString = getAttribute(SVGNames::hrefAttr, XLinkNames::hrefAttr);
-    if (!location.specifiedURLString.isNull())
-        location.resolvedURL = document().completeURL(location.specifiedURLString);
+    auto location = CSS::completeURL(getAttribute(SVGNames::hrefAttr, XLinkNames::hrefAttr), document()).value_or(CSS::URL::empty());
```
Looks like here my annoyance at the name `url` caused me to use the word `location`, not `locator`, nor `srcURL`. 😂

darinadler reviewed (approved):
This looks right; clearly we are relying on tests to be sure, since there are minor semantic changes to different code paths as we unify our approach.

weinig committed:
Convert URL types to strongly typed CSS/Style values
https://bugs.webkit.org/show_bug.cgi?id=291045

Reviewed by Antti Koivisto and Darin Adler.

Update parsing and representation of <url> productions as
defined in CSS Values and Units 4 (https://www.w3.org/TR/css-values-4/#urls)
to use strong value types and have consistent behavior for
serialization and style building. This fixes a few WPT test
failures.

As URLs are no longer representable by CSSPrimitiveValue,
a new CSSURLValue type is also added that wraps CSS::URL.

This change allows removing a lot of unnecessary re-resolution
that was going on in the image code, as the only re-resolution
needed can be performed by style building. Ultimately, it would
be better if we could remove that final case as well, by finding
a good hook to re-parse necessary styles adopted from documents,
but that can be done separately.

To allow using Style::URL in ReferenceFilterOperation without a
laying violation (or rather, without additional layering violations
as it was already using non-platform types), ReferenceFilterOperation
was moved out of platform. This matches what was done recently for
DropShadowFilterOperationWithStyleColor.

* LayoutTests/fast/masking/parsing-clip-path-iri-expected.txt:
* LayoutTests/fast/masking/parsing-clip-path-iri.html:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/fill-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/stroke-computed-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSCursorImageValue.cpp:
* Source/WebCore/css/CSSCursorImageValue.h:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSFilterPropertyValue.cpp:
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
* Source/WebCore/css/CSSFontFaceSrcValue.h:
* Source/WebCore/css/CSSImageValue.cpp:
* Source/WebCore/css/CSSImageValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSURLValue.cpp: Added.
* Source/WebCore/css/CSSURLValue.h: Added.
* Source/WebCore/css/CSSUnits.cpp:
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.cpp:
* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
* Source/WebCore/css/calc/CSSCalcTree+NumericIdentity.h:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcType.cpp:
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Masking.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Motion.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.h:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/typedom/CSSStyleImageValue.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/values/CSSValueTypes.cpp:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/filter-effects/CSSFilterReference.h:
* Source/WebCore/css/values/primitives/CSSURL.cpp: Added.
* Source/WebCore/css/values/primitives/CSSURL.h: Added.
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
* Source/WebCore/page/ElementTargetingController.cpp:
* Source/WebCore/platform/graphics/filters/FilterOperation.cpp:
* Source/WebCore/platform/graphics/filters/FilterOperation.h:
* Source/WebCore/rendering/CSSFilter.cpp:
* Source/WebCore/rendering/PathOperation.cpp:
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/ReferencedSVGResources.h:
* Source/WebCore/rendering/RenderLayerFilters.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/style/ReferenceFilterOperation.cpp: Added.
* Source/WebCore/rendering/style/ReferenceFilterOperation.h: Added.
* Source/WebCore/rendering/style/SVGRenderStyle.h:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.h:
* Source/WebCore/rendering/style/StyleCachedImage.cpp:
* Source/WebCore/rendering/style/StyleCachedImage.h:
* Source/WebCore/rendering/style/StyleCursorImage.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.h:
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
* Source/WebCore/rendering/style/StyleImage.h:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResources.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/filter-effects/StyleFilterProperty.cpp:
* Source/WebCore/style/values/filter-effects/StyleFilterReference.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleFilterReference.h: Added.
* Source/WebCore/style/values/primitives/StyleURL.cpp: Added.
* Source/WebCore/style/values/primitives/StyleURL.h: Added.
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.cpp:
* Source/WebCore/svg/SVGFontFaceUriElement.cpp:
* Source/WebCore/svg/SVGURIReference.cpp:
* Source/WebCore/svg/SVGURIReference.h:
* Source/WebCore/svg/graphics/SVGResourceImage.cpp:
* Source/WebCore/svg/graphics/SVGResourceImage.h:

Canonical link: https://commits.webkit.org/293295@main
